cmd/gateway/hdfs/gateway-hdfs.go

```diff
@@ -0,0 +1,706 @@
+	_, err = n.clnt.Stat(minio.PathJoin(hdfsSeparator, bucket))
+	if err != nil {
+		return info, hdfsToObjectErr(ctx, err, bucket)
+	}
+
+	var w *hdfs.FileWriter
+	w, err = n.clnt.Append(minio.PathJoin(hdfsSeparator, minioMetaTmpBucket, uploadID))
+	if err != nil {
+		return info, hdfsToObjectErr(ctx, err, bucket, object, uploadID)
+	}
+
+	_, err = io.Copy(w, r.Reader)
+	if err != nil {
+		w.Close()
+		return info, hdfsToObjectErr(ctx, err, bucket, object, uploadID)
+	}
+	w.Close()
+
+	info.PartNumber = partID
+	info.ETag = r.MD5CurrentHexString()
+	info.LastModified = minio.UTCNow()
+	info.Size = r.Reader.Size()
+
+	return info, nil
+}
+
+func (n *hdfsObjects) CompleteMultipartUpload(ctx context.Context, bucket, object, uploadID string, parts []minio.CompletePart, opts minio.ObjectOptions) (objInfo minio.ObjectInfo, err error) {
+	_, err = n.clnt.Stat(minio.PathJoin(hdfsSeparator, bucket))
+	if err != nil {
+		return objInfo, hdfsToObjectErr(ctx, err, bucket)
+	}
+
+	if err = n.checkUploadIDExists(ctx, bucket, object, uploadID); err != nil {
+		return objInfo, err
+	}
+
+	name := minio.PathJoin(hdfsSeparator, bucket, object)
+	dir := path.Dir(name)
+	if dir != "" {
+		if err = n.clnt.MkdirAll(dir, os.FileMode(0755)); err != nil {
+			return objInfo, hdfsToObjectErr(ctx, err, bucket, object)
+		}
+	}
+
+	err = n.clnt.Rename(minio.PathJoin(hdfsSeparator, minioMetaTmpBucket, uploadID), name)
+	// Object already exists is an error on HDFS
+	// remove it and then create it again.
+	if os.IsExist(err) {
+		if err = n.clnt.Remove(name); err != nil {
+			if dir != "" {
+				n.deleteObject(minio.PathJoin(hdfsSeparator, bucket), dir)
+			}
+			return objInfo, hdfsToObjectErr(ctx, err, bucket, object)
+		}
+		if err = n.clnt.Rename(minio.PathJoin(hdfsSeparator, minioMetaTmpBucket, uploadID), name); err != nil {
+			if dir != "" {
+				n.deleteObject(minio.PathJoin(hdfsSeparator, bucket), dir)
+			}
+			return objInfo, hdfsToObjectErr(ctx, err, bucket, object)
+		}
+	}
+	fi, err := n.clnt.Stat(name)
+	if err != nil {
+		return objInfo, hdfsToObjectErr(ctx, err, bucket, object)
+	}
+
+	// Calculate s3 compatible md5sum for complete multipart.
+	s3MD5, err := minio.GetCompleteMultipartMD5(ctx, parts)
+	if err != nil {
+		return objInfo, err
+	}
+
+	return minio.ObjectInfo{
+		Bucket:  bucket,
+		Name:    object,
+		ETag:    s3MD5,
+		ModTime: fi.ModTime(),
+		Size:    fi.Size(),
+		IsDir:   fi.IsDir(),
+		AccTime: fi.(*hdfs.FileInfo).AccessTime(),
+	}, nil
+}
+
+func (n *hdfsObjects) AbortMultipartUpload(ctx context.Context, bucket, object, uploadID string) (err error) {
+	_, err = n.clnt.Stat(minio.PathJoin(hdfsSeparator, bucket))
+	if err != nil {
+		return hdfsToObjectErr(ctx, err, bucket)
+	}
+	return hdfsToObjectErr(ctx, n.clnt.Remove(minio.PathJoin(hdfsSeparator, uploadID)), bucket, object, uploadID)
+}
```

Comment:
AbortMultipartUploads is not finding upload id, n.clnt.Remove needs to have minioMetaTmpBucket in the path
```
➜  hadoop aws s3api create-multipart-upload   --bucket tbucket11 --key etag2 --endpoint-url http://localhost:9000
{
    "Bucket": "tbucket11", 
    "UploadId": "4898aeb4-3a6b-40ba-9650-6160e7acc3c9", 
    "Key": "etag2"
}
➜  hadoop aws s3api abort-multipart-upload --upload-id 4898aeb4-3a6b-40ba-9650-6160e7acc3c9   --bucket tbucket11 --key etag2   --endpoint-url http://localhost:9000  --no-verify-ssl
An error occurred (NoSuchUpload) when calling the AbortMultipartUpload operation: The specified multipart upload does not exist. The upload ID may be invalid, or the upload may have been aborted or completed.

```